crates/sui-framework/packages/sui-framework/sources/account.move

```diff
@@ -0,0 +1,112 @@
+    let key = AccountKey { address, ty };
+    return field::hash_type_and_key(sui_account_root_address(), key)
+}
+
+public struct Reservation<phantom T> has drop {
+    owner: address,
+    limit: u64,
+}
+
+fun decrement<T>(self: &mut Reservation<T>, amount: u64) {
+    assert!(self.limit >= amount, EInsufficientFunds);
+    self.limit = self.limit - amount;
+}
+
+/// The scheduling/execution layer ensures that a reserve() call is never permitted unless
+/// there are sufficient funds. `Reservation<T>` thus serves as a proof that the account
+/// cannot be overdrawn.
+entry fun reserve<T>(owner: address, limit: u64, ctx: &TxContext): Reservation<T> {
+    // TODO: handle sponsored transactions and (in the future) multi-agent transactions
+    assert!(ctx.sender() == owner, ENotAccountOwner);
+    return Reservation { owner, limit }
+}
+
+/// Withdraw from an account.
+/// Requires a reservation of the appropriate type, with proof of enough funds to cover the
+/// withdrawal.
+///
+/// `value` is an amount to be debited from the account. Therefore, modules who wish to
+/// ensure conservation should use the following pattern, in which identical and offsetting
+/// credits and debits are created at the same time:
+/// 
+/// fun withdraw(reservation: &mut Reservation<Foo>, amount: u64, ctx: &TxContext): Foo {
+///   let debit = MergableFoo { value: amount };
+///   let credit = Foo { value: amount };
+///   account::withdraw_from_account(reservation, debit, amount, ctx.sender());
+///   credit
+/// }
+public fun withdraw_from_account<T>(
+    // reservation proves that the account has enough funds, and that the withdrawal
+    // is authorized
+    reservation: &mut Reservation<T>, 
+    // debit is a typed wrapper around `amount`. It must contain the same value
+    // as is stored in `amount`. iiuc we should be able to remove this duplication when
+    // signatures are available.
+    debit: T,
+    amount: u64,
+) {
+    // Conservation: aborts if reservation is insufficient
+    reservation.decrement(amount);
+    let account_address = get_account_field_address<T>(reservation.owner);
+    // Conservation:
+    // - `debit` will be subtracted from the account
+    // - No new reservations will be issued without taking into account the debit.
+    split_from_account(debit, account_address);
+}
+
+/// Transfer a value to an account.
+///
+/// TODO: requires move verifier changes (analagous to the `transfer` checks) that ensures that
+/// this can only be called from the module in which T is defined. Modules can implement secure
+/// accounts without this by using a private type that cannot be constructed outside the module.
+///
+/// Because types must explicitly implement a conversion from their ordinary type to a mergable
+/// type (i.e. one made only of types defined in mergable.move), there is no need for an analogue
+/// to `public_transfer`.
+public fun transfer_to_account<T>(deposit: T, recipient: address) {
```

Comment:
Was thinking about this, and I unfortunately don't think it's safe if:
1. `recipient` is an object ID (rather than an end-user address)
2. We allow transfers to "wrapped" accounts (e.g., `struct Obj has key { .., wrapped_account_id: ID }`)

The reason is that with wrapped `Balance`/`Coin` fields, you often want to maintain invariants that could be violated by a direct transfer. For example, in a classic CFMM pool

```
struct Pool has key {
  // ...
  coin1_account_id: ID, // before: Balance<T1>
  coin2_account_id: ID, // before: Balance<T2>
}
```

, you would want to maintain the invariant `balance(coin1_account_id) = k * balance(coin2_account_id)`. With the current `Balance`-based approach, you leverage encapsulation to enforce this--e.g. only allow writes in `swap`, and only if the invariant are maintained. But with an implementation that used wrapped accounts, it would not be possible to enforce this invariant because of (2) (e.g., can just deposit directly to balance 1 without touching balance 2).

Reply:
Two approaches come to mind for addressing this--the key ideas behind both are to enforce encapsulation on transfers to "object accounts":
A. `transfer_to_account` requires a `&mut UID` or `&ID` of object `O` to transfer to an object account for `O` (not needed for an address account, of course)
B. Object accounts must be explicitly created, and have a different/more structured API for transfers than addresses. E.g.,

```
struct Account has store {
  id: UID,
}

public fun create(ctx: &mut TxContext): Account

// could also split into more fine-grained caps
public fun transfer_to_account<T>(deposit: T, account: &mut Account)
```